agent/agentcontainers/api.go

```diff
@@ -263,10 +240,102 @@ func (api *API) loop() {
 	}
 }
 
+// updaterLoop is responsible for periodically updating the container
+// list and handling manual refresh requests.
+func (api *API) updaterLoop() {
+	defer close(api.updaterDone)
+	defer api.logger.Debug(api.ctx, "updater loop stopped")
+	api.logger.Debug(api.ctx, "updater loop started")
+
+	// Ensure that only once instance of the updateContainers is running
+	// at a time. This is a workaround since quartz.Ticker does not
+	// allow us to know if the routine has completed.
+	sema := make(chan struct{}, 1)
+	sema <- struct{}{}
+
+	// Ensure only one updateContainers is running at a time, others are
+	// queued.
+	doUpdate := func() error {
+		select {
+		case <-api.ctx.Done():
+			return api.ctx.Err()
+		case <-sema:
+		}
+		defer func() { sema <- struct{}{} }()
+
+		return api.updateContainers(api.ctx)
```

Comment:
Would it make sense instead to just return early from `doUpdate()` if we're not able to read from `sema`? Queueing multiple updates seems like it could cause increased memory usage over time, and I'm not sure what value we get from running it right again after one instance has just completed.

Reply:
If we did that, we can't wait for it to complete in the `refreshContainers` function. I'll see if there's any point in keeping a refresh method around after I refactor the recreate endpoint.